log/log.go

```diff
@@ -14,34 +16,98 @@ import (
 var (
 	log      *zap.SugaredLogger
 	errorLog *os.File
+	output   string
+	logLevel sync.Map // this is a thread-safe map[string]string
+)
+
+const (
+	defaultLevel  = "*:error"
+	defaultOutput = "stderr"
 )
 
 func init() {
 	// Allow overriding the default log level via $LOG_LEVEL, so that the
 	// environment variable can be set globally even when running tests.
 	// Always initializing the logger is also useful to avoid panics when
 	// logging if the logger is nil.
-	level := "error"
+	level := []string{defaultLevel}
 	if s := os.Getenv("LOG_LEVEL"); s != "" {
-		level = s
+		level = strings.Split(s, ",")
 	}
-	Init(level, "stderr")
+
+	InitWithLevels(level, defaultOutput)
+}
+
+// Init initializes the logger, with a single logLevel (debug|info|warn|error|fatal).
+// out can be either "stdout", "stderr" or "/path/to/file" and it's a global setting.
+//
+// kept for backwards compatibility, it just wraps InitWithLevels()
+func Init(logLevel string, out string) {
+	InitWithLevels([]string{logLevel}, out)
 }
 
-func Logger() *zap.SugaredLogger { return log }
+// InitWithLevels initializes the logger.
+// logLevels are strings like "somename:debug" or "*:info" (equivalent to simply "info").
+// out can be either "stdout", "stderr" or "/path/to/file" and it's a global setting.
+//
+// It's safe to use concurrently.
+func InitWithLevels(logLevels []string, out string) {
+	for _, ll := range logLevels {
+		split := strings.Split(ll, ":")
+		if len(split) > 1 {
+			// split[0] is subsystem name, split[1] is level
+			logLevel.Store(split[0], split[1])
```

Comment:
if you set up all named loggers as static globals, then in here you can also check if a log level component is unknown; otherwise we'll silently ignore e.g. `foobar:info`.